### ISSUE TITLE
Client options not updated such as fetchOptions

### ISSUE THREAD

**Issue by addstar34** [FormidableLabs/next-urql#16]
If you visit a page using `withUrqlClient` HOC and then login and visit another page and want to set an authorization header like the following (from the readme) the client isn't reinitialized and thus the authorization header is not set.

```
withUrqlClient(ctx => ({
  url: 'http://localhost:3000',
  fetchOptions: {
    headers: {
      Authorization: `Bearer ${ctx.req.headers.token}`,
    },
  },
}));
```

**Comment by parkerziegler** [comment]
Interesting, can you give a little more information about your setup? Are you saying that you have something like the following architecture:

```
pages
|__login.js --> wrapped with withUrqlClient
|__other_page.js --> not wrapped with withUrqlClient
```

And that, when navigating to `other_page.js`, `withUrqlClient` never gets re-run? I would mark that as expected behavior if so, since `withUrqlClient` is only wrapping `login.js` and should only get run when that page is rendered.

If you need an `urqlClient` instance on `other_page.js`, you'll need to also wrap that in `withUrqlClient`, or opt for using `_app.js`. If you can give me more specific details on your architecture I can probably help further. We've been rather tightly focused with the lib on just getting queries prefetched on the server, so I wouldn't be surprised if there are edge cases we missed. As much information as you can provide will help us better provide advice. Thanks!

**Comment by addstar34** [comment]
I have the following pages:

```
pages\users.js --> wrapped with withUrqlCLient
pages\sign-in.js --> not wrapped
```

When I wrap the pages `withUrqlClient` I check for an in memory token and want to set that as the authorizaton header eg:

```
\\ pages\users.js

export default withUrqlClient(ctx => {
  const token = getToken() // I store an access token in memory

  return {
    url: 'http://localhost:3000',
    fetchOptions: {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    },
  }
})(UsersPage)
```

So what happens if if I'm not logged in and I got to the users page then `withUrqlClient` will init the urql client with an authorization header that has no token as I haven't signed in yet. Then if I go to the sign in page and then revisit the users page it doesn't update the fetchOptions.

I believe the issue is because in `init-urql-client.ts` you have this:

```
if (isServer || !urqlClient) {
    ssrCache = ssrExchange({ initialState });

    urqlClient = createClient({
      ...clientOptions,
      suspense: isServer,
      exchanges: mergeExchanges(ssrCache),
    });
  }
```

so even though `with-urql-client.tsx` uses React.useMemo with dependencies of `[urqlClient, pageProps, clientOptions, urqlState]` it still only calls `init-urql-client.ts` which only checks if its on the server or if there's no urqlClient and because we already have a urqlClient then it doesn't recreate the client with the new fetchOptions.

Let me know if this makes sense or not, I could possibly push up a modified version of my project that shows the issue if you need.

**Comment by parkerziegler** [comment]
Ah I see what you're saying, and that does make sense. For the first render of `pages\users.js`, `withUrqlClient` will get run on the server, setting up a Client instance with some fetch options (in this case, an `undefined` token, since the user isn't logged in). Then, when navigating back to `pages\users.js`, `withUrqlClient` (and its `getInitialProps`) will get run on the client-side only and, since there's an extant Client instance, we use it and we won't actually apply the `token`.

That's an interesting problem, and I'll have to think some more on how to solve it. I think you're right that it has to do with when and under what conditions `initUrqlClient` gets called and perhaps also the fact that `urqlClient` referenced in that file is more or less a singleton instance. I have a setup locally that I think mirrors your issue pretty well, so I can debug more from here. Thanks for opening the issue, I'm not certain about when we'll get a fix in, but I'll try to rope some fellow Formidables in.

**Comment by addstar34** [comment]
I’ve worked around this by setting the authorization header via a custom exchange and using `withUrqlClient` HOC on `_app.js`.

I wasn’t able to use the next-urql package as I needed to do a few extra things so I just copied it into my project. I’m not sure if its useful or not to know but what I had to change was:
- I needed to pass nextjs’s context object into the exchanges so I could set an access token on it so after server rendering it can get re-hydrated client side.
- I also need to call my refresh token endpoint after `ssrPrepass` in case it was rendering a page without any graphql queries. This was also to set the access token so it could get re-hydrated.
- I re-hydrate the access token in the withUrql component before it memo’s the client.

My refresh token exchange was based off your custom exchange here https://formidable.com/open-source/urql/docs/guides/#authentication
So I’ve very grateful for these docs, thank you!

So in saying that I’m not sure if this is an issue or if the way to set/update `fetchOptions` should be done via a custom exchange?

**Comment by kvnwolf** [comment]
Ha! I was just going to report that. I'd even create a sandbox demonstrating it: https://codesandbox.io/s/next-urql-client-refresh-error-uhed5

If you are at `/` and move to `/other`, the client doesn't get updated, only if you refresh the page.

**Comment by parkerziegler** [comment]
👋 Hey folks, thanks for your patience here, I'm doing a little more digging on this today. The key piece for consideration is that we can't just re-create the Client on the client-side for every page navigation – if we did, the new client would have no cache results to rehydrate from.

I'm considering adding some sort of functional addition to the API, like an `updateClient` callback that would get run on every page transition where you can specify how you need the Client updated on a per page basis. I'm not sure how kosher this is, so also going to tag @kitten and @JoviDeCroock for their ideas here 😅

**Comment by nghiepdev** [comment]
I have the same issue.

```js
export default withUrqlClient(
  ctx => {
    return {
      url: process.browser ? 'api-client' : 'api-server',
    };
  }
)(MyApp);
```
The url alway is `api-server`.

**Comment by parkerziegler** [comment]
@nghiepit the case you outline above isn't really one `next-urql` is designed to support. There's not really a point in SSR'ing your GraphQL requests if you're going to be making requests to an entirely different endpoint on the client side 😅 Remember, the core premise of something like `next-urql` is that your GraphQL queries get executed server-side _and then_ your components get hydrated client-side with the data resolved in those server-side requests. If you're requesting different endpoints on the client vs. the server, then you'd likely need to wire up a [custom exchange](https://formidable.com/open-source/urql/docs/guides/) to handle directing all client-side requests to one url and all server-side requests to another.

We're working on getting an example showing a custom exchange soon to help alleviate confusion around this question. Ultimately, the `Client` instance created on the server shouldn't really change in the course of your application's lifecycle, and exchanges give a semantic path to allowing you to add things like a client-side derived auth token to all future requests _after_ the `Client` has been initialized. But, again, if we recreated a new `Client` instance on every page navigation, we'd basically never get any cached results and lose the vast majority of benefit that a GraphQL client like `urql` provides.

**Comment by nghiepdev** [comment]
@parkerziegler  thank you for response.

> There's not really a point in SSR'ing your GraphQL requests if you're going to be making requests to an entirely different endpoint on the client side 😅 Remember, the core premise of something like next-urql is that your GraphQL queries get executed server-side and then your components get hydrated client-side with the data resolved in those server-side requests.

The reason to have such a request different endpoints on the client vs. the server because:

> **GraphQL Endpoint:** http://localhost:4000 -> Nginx proxy_pass -> Public API-> https://mydomain-graphql.com

So, on the server If I request directly to http://localhost:4000 will be faster.

In the real project, Usually , deploy to Docker, the containers are same networks, this is very helpful for us.

p/s: It would be great if you could help to explain how to **custom exchange** to make this.

**Comment by kitten** [comment]
@nghiepit To me your code looks fine. What @parkerziegler was hinting at is that with SSR you may not see the client execute any queries on load on its own. Instead it will just use the results that the server previously requested during SSR.

So what you should see is, if you then navigate to a different page, the client will fetch your queries (hopefully with the correct URL), since a client-side navigation doesn’t involve an SSR/server round trip

**Comment by parkerziegler** [comment]
> So what you should see is, if you then navigate to a different page, the client will fetch your queries (hopefully with the correct URL), since a client-side navigation doesn’t involve an SSR/server round trip.

@kitten this is the tricky bit. Basically, the way things are working right now is that we create the client **once** on the server and reuse it on the client-side to access its cached results. This means that, if you attempt to change the client's configuration **on the client-side** after that initial render, you won't actually see it updated (it's still the reused instance from the server).

My idea around how to fix this is to provide an example of a custom exchange, where you modify an outgoing **operation** to use a different `url` that than the one supplied by the client. Inside that exchange you should have all the info you need to determine if you're running on the server or the client, and can alter your operations as needed.

Let me know if you think this is a reasonable approach to direct users towards, or if you think we need to come up with a better strategy for allowing client-side alterations to the client instance.

**Comment by parkerziegler** [comment]
This has been fixed and should be available in v0.3.0 and later of `next-urql`. If you're still seeing issues, please open a new issue over in the [`urql` repo](https://github.com/FormidableLabs/urql), as `next-urql` has moved there.
